sys/dev/isp/isp.c

```diff
@@ -4322,18 +4340,362 @@ isp_reinit(ispsoftc_t *isp, int do_load_defaults)
+				fcp->flt_region_nvram = 0;
+			}
+			fcp->flt_region_nvram += ISP2400_NVRAM_PORT_ADDR(isp->isp_port);
+		}
+	} else {
+		fcp->flash_data_addr = 0x7fe7c000;
+		fcp->flt_region_nvram = 0;
+		fcp->flt_region_nvram += ISP2400_NVRAM_PORT_ADDR(isp->isp_port);
+	}
 	return (isp_read_nvram_2400(isp));
 }
 
+static int
+isp_read_flthdr_2xxx(ispsoftc_t *isp)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int retval = 0;
+	uint32_t addr, lwrds, *dptr;
+	uint16_t csum;
+	uint8_t flthdr_data[FLT_HEADER_SIZE];
+
+	addr = fcp->flt_region_flt;
+	dptr = (uint32_t *) flthdr_data;
+
+	isp_prt(isp, ISP_LOGDEBUG0,
+	    "FLTL[DEF]: 0x%x", addr);
+	for (lwrds = 0; lwrds < FLT_HEADER_SIZE >> 2; lwrds++) {
+		isp_rd_2xxx_flthdr(isp, addr++, dptr++);
+	}
+	dptr = (uint32_t *) flthdr_data;
+	for (csum = 0, lwrds = 0; lwrds < FLT_HEADER_SIZE >> 4; lwrds++) {
+		uint16_t tmp;
+		ISP_IOXGET_16(isp, &dptr[lwrds], tmp);
+		csum += tmp;
+	}
+	if (csum != 0) {
+		retval = -1;
+		goto out;
+	}
+	isp_parse_flthdr_2xxx(isp, flthdr_data);
+out:
+	return (retval);
+}
+
+static void
+isp_rd_2xxx_flthdr(ispsoftc_t *isp, uint32_t addr, uint32_t *rp)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int loops = 0;
+	uint32_t base = fcp->flash_data_addr;
+	uint32_t tmp = 0;
+
+	ISP_WRITE(isp, BIU2400_FLASH_ADDR, base | addr);
+	for (loops = 0; loops < 5000; loops++) {
+		ISP_DELAY(10);
+		tmp = ISP_READ(isp, BIU2400_FLASH_ADDR);
+		if ((tmp & (1U << 31)) != 0) {
+			break;
+		}
+	}
+	if (tmp & (1U << 31)) {
+		*rp = ISP_READ(isp, BIU2400_FLASH_DATA);
+		ISP_SWIZZLE_NVRAM_LONG(isp, rp);
+	} else {
+		*rp = 0xffffffff;
+	}
+}
+
+static void
+isp_parse_flthdr_2xxx(ispsoftc_t *isp, uint8_t *flthdr_data)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	uint16_t ver, csum;
+
+	ver = le16toh((uint16_t) (ISP2XXX_FLT_VERSION(flthdr_data)));
+	fcp->flt_length = le16toh((uint16_t) (ISP2XXX_FLT_LENGTH(flthdr_data)));
+	csum = le16toh((uint16_t) (ISP2XXX_FLT_CSUM(flthdr_data)));
+
+	if ((fcp->flt_length == 0 ) ||
+	    (fcp->flt_length > (FLT_HEADER_SIZE + FLT_REGIONS_SIZE))) {
+		isp_prt(isp, ISP_LOGERR,
+		    "FLT[DEF]: Invalid length=0x%x(%d)",
+		    fcp->flt_length, fcp->flt_length);
+	}
+	isp_prt(isp, ISP_LOGCONFIG,
+	    "FLT[DEF]: version=0x%x length=0x%x(%d) checksum=0x%x",
+	    ver, fcp->flt_length, fcp->flt_length, csum);
+}
+
+static int
+isp_read_flt_2xxx(ispsoftc_t *isp)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int retval = 0;
+	int len = fcp->flt_length - FLT_HEADER_SIZE;
+	uint32_t addr, lwrds, *dptr;
+	uint8_t flt_data[len];
+	fcp->flt_region_entries = len / FLT_REGION_SIZE;
+
+	addr = fcp->flt_region_flt + (FLT_HEADER_SIZE >> 2);
+	dptr = (uint32_t *) flt_data;
+	isp_prt(isp, ISP_LOGCONFIG, "FLT[DEF]: regions=%d",
+	    fcp->flt_region_entries);
+	for (lwrds = 0; lwrds < len >> 2; lwrds++) {
+		isp_rd_2xxx_flt(isp, addr++, dptr++);
+	}
+	retval = isp_parse_flt_2xxx(isp, flt_data);
+	return (retval);
+}
+
+static void
+isp_rd_2xxx_flt(ispsoftc_t *isp, uint32_t addr, uint32_t *rp)
```

Comment:
Is this identical copy of isp_rd_2xxx_flthdr()? Third copy of isp_rd_2400_nvram()?

Reply:
Fixed now, there is only one function left to do the real read from flash